Microsoft.Toolkit.Uwp.UI.Controls/DockPanel/DockPanel.cs

```diff
@@ -0,0 +1,135 @@
+        }
+
+        /// <inheritdoc />
+        protected override Size ArrangeOverride(Size finalSize)
+        {
+            if (Children.Count == 0)
+            {
+                return finalSize;
+            }
+
+            var currentSize = new Rect(0, 0, finalSize.Width, finalSize.Height);
+            var childrenCount = LastChildFill ? Children.Count - 1 : Children.Count;
+
+            for (var index = 0; index < childrenCount; index++)
+            {
+                var child = Children[index];
+                child.SetValue(Canvas.ZIndexProperty, Children.Count - index);
+                var dock = (Dock)child.GetValue(DockProperty);
+                double width, height;
+                switch (dock)
+                {
+                    case Dock.Left:
+
+                        width = child.DesiredSize.Width;
+                        child.Arrange(new Rect(currentSize.X, currentSize.Y, width, currentSize.Height - currentSize.Y));
+                        currentSize.X += width;
+
+                        break;
+                    case Dock.Top:
+
+                        height = child.DesiredSize.Height;
+                        child.Arrange(new Rect(currentSize.X, currentSize.Y, currentSize.Width - currentSize.X, height));
+                        currentSize.Y += height;
+
+                        break;
+                    case Dock.Right:
+
+                        width = child.DesiredSize.Width;
+                        child.Arrange(new Rect(currentSize.Width - width, currentSize.Y, width, currentSize.Height - currentSize.Y));
+                        currentSize.Width -= width;
+
+                        break;
+                    case Dock.Bottom:
+
+                        height = child.DesiredSize.Height;
+                        child.Arrange(new Rect(currentSize.X, currentSize.Height - height, currentSize.Width - currentSize.X, height));
+                        currentSize.Height -= height;
+
+                        break;
+                }
+            }
+
+            if (LastChildFill)
+            {
+                var width = currentSize.Width - currentSize.X;
+                var height = currentSize.Height - currentSize.Y;
+                var child = Children[Children.Count - 1];
+                child.Arrange(
+                    new Rect(currentSize.X, currentSize.Y, width > 0 ? width : 0, height > 0 ? height : 0));
+                child.SetValue(Canvas.ZIndexProperty, 0);
+            }
+
+            return finalSize;
+        }
+
+        /// <inheritdoc />
+        protected override Size MeasureOverride(Size availableSize)
+        {
+            var width = availableSize.Width;
+            var height = availableSize.Height;
+
+            if (double.IsInfinity(width))
+            {
+                width = Window.Current.Bounds.Width;
+            }
+
+            if (double.IsInfinity(height))
+            {
+                height = Window.Current.Bounds.Height;
```

Comment:
The max size isn't the window bounds. Those bounds could be too large or too small. Infinity could for instance happen inside a scroll viewer. You should let the infinite values pass on to the Measure call on the children. That just gives you a desired size that might be larger than what you have room for, but that's ok. You can then limit them further during arrange.
The returned finalSize should then return the max-sizes combined based on docking locations. Ie width would be LeftDock desired width + center fill desired width + right dock desired width. (finalSize can never return infinite values, since this becomes the desired size of the dock panel - during arrange the container might give you less space)

Reply:
if we didn't limit the bounds we will be having a problem which is, I don't know the (center fill desired width) I only know the LeftDock and RightDock :) the DockPanel should calculate the center by it self.
for example I have a infinity width and I have one dock right and one left, how can I measure the center :)
That's why I see we need to limit

Reply:
Just a suggestion, can we use the combined width/height of the children to specify the total size instead of the window size (if the size is infinity). For example, when the width is infinity, calculate the final width by calculating the sum of the desired width for all children requested to be left or right.

Reply:
For the sake of argument, could we not use the same code the WPF dock panel uses in these cases?

Reply:
There is no fun in copying and pasting, plus I believe that the toolkit team can deliver same results as WPF if not better 😝.
I know there would also be problems related to licensing in reusing WPF soruce code, you can get back to Nikola in the licensing area if you want.

Reply:
Well, we don't have to copy the code wholesale, but they do have working logic you could base your approach on for this particular issue :P Thought it's not too dissimilar from @nmetulev 's suggestion at any rate if you'd rather avoid peaking

Reply:
AFAIK, the reference source is only meant to be used as a reference, don't believe the license allows us to copy it and use it here.